apps/system/js/notifications_system_message.js

```diff
@@ -0,0 +1,157 @@
+/* globals BaseModule,
+           Notification
+*/
+
+'use strict';
+
+(function(exports) {
+  var NotificationsSystemMessage = function() {};
+
+  NotificationsSystemMessage.SERVICES = [
+    'handleSystemMessageNotification',
+    'unhandleSystemMessageNotification'
+  ];
+
+  BaseModule.create(NotificationsSystemMessage, {
+    name: 'NotificationsSystemMessage',
+    EVENT_PREFIX: 'systemMessageNotification',
+    DEBUG: false,
+    TRACE: false,
+
+    _handlers: {},
+
+    _start: function nsm_start() {
+      this.registerHandler();
+    },
+
+    registerHandler: function nsm_registerHandler() {
+      this.debug('Installing system app notification message handler');
+      navigator.mozSetMessageHandler('notification',
+        this.processSystemMessage.bind(this));
+    },
+
+    processSystemMessage: function(message) {
+      this.debug('Received system message notification: ' +
+        JSON.stringify(message));
+
+      if (!message.clicked) {
+        return;
+      }
+
+      if (!message.data || !('systemMessageTarget' in message.data)) {
+        console.warn('No data payload or system message target defined, ' +
+                     'considering legacy and force-closing.');
+        this.closeOldNotification(message);
+        return;
+      }
+
+      var target = message.data.systemMessageTarget;
+      if (!target) {
+        console.error('System message target seems undefined, force-closing');
+        this.closeOldNotification(message);
+        return;
+      }
+
+      if (!this.hasRegisteredTarget(target)) {
+        console.error('No module has subscribed to: ' + target);
+        return;
+      }
+
+      this.callRegisteredTarget(target, message);
+    },
+
+    registerTarget: function(target, obj) {
+      if (this.hasRegisteredTarget(target)) {
+        this.debug('Already have a handler for target: ' + target);
+        return;
+      }
+
+      this.debug('Registering a handler for target: ' + target);
+      this._handlers[target] = {
+        name: target,
+        context: obj
+      };
+    },
+
+    unregisterTarget: function(target, obj) {
+      if (!this.hasRegisteredTarget(target)) {
+        console.error('No handler to unregister for target: ' + target);
+        return;
+      }
+
+      var c = this.getTargetContext(target);
+      if (c !== obj) {
+        console.error('Mismatching object to unregister for target: ' + target);
+        return;
+      }
+
+      this.debug('Unregistering the handler for target: ' + target);
+      this._handlers[target] = null;
+      delete this._handlers[target];
+    },
+
+    callRegisteredTarget: function(target, payload) {
+      var ctx = this.getTargetContext(target);
+      if (!ctx) {
+        console.error('No registered target for ' + target);
+        return;
+      }
+
+      ctx.handleSystemMessageNotification.call(ctx, payload);
```

Comment:
Let's make sure ctx has the `handleSystemMessageNotification` method defined and if not console.error the problem.